### DISCUSSION TITLE
Built-in way to coerce null to undefined

### DISCUSSION THREAD

**Discussion by bluepnume** [colinhacks/zod#2103] [Q&A] [answered]
Right now I'm doing something like this:

```typescript
const coerceNullToUndefined = <Type>(value : Type | null | undefined) : Type | undefined => {
    return value === null ? undefined : value;
};

const apiResponseSchema = z.object({
    name: z.string().nullish().transform(coerceNullToUndefined)
});
```

That makes the resulting value a little more sane; I don't have to worry about dealing with both null and undefined.

It would be awesome if it was possible to have a built-in way to do this, like:

```typescript
const apiResponseSchema = z.object({
    name: z.string().nullish().coerceNullToUndefined()
});
```

**Comment by JacobWeisenburger** [comment, marked as answer]
A little cleaner way to do this is:

```ts
const schema = z.string().nullish().transform( x => x ?? undefined )
type Data = z.infer<typeof schema>
// type Data = string | undefined

console.log( schema.parse( 'foo' ) ) // 'foo'
console.log( schema.parse( null ) ) // undefined
console.log( schema.parse( undefined ) ) // undefined
```

Because this is quite a simple and clean way to do this. I'm not sure we need a built-in way to do this. Thoughts?

**Reply by sschneider-ihre-pvs** [reply]
But what if you do something like `z.coerce.date()` with a null value? you would get `1970-01-01T00:00:00.000Z`. how do you filter that out?

**Reply by JacobWeisenburger** [reply]
Is this what you are looking for?

```ts
const schema = z.union( [
    z.string(),
    z.number(),
    z.date(),
    z.null().transform( x => x ?? undefined ),
] ).pipe( z.coerce.date() )
type Data = z.infer<typeof schema>
// type Data = Date

console.log( schema.parse( '2024-03-05T00:00:00.000Z' ) ) // 2024-03-05T00:00:00.000Z
console.log( schema.parse( 1709596800000 ) )              // 2024-03-05T00:00:00.000Z
console.log( schema.parse( new Date( 1709596800000 ) ) )  // 2024-03-05T00:00:00.000Z

const result = schema.safeParse( null )
!result.success && console.log( result.error.issues )
// [
//     {
//         code: "invalid_date",
//         path: [],
//         message: "Invalid date",
//     }
// ]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by INDAPP** [reply]
What if I need to use this solution in the following situation?

```ts
const schema = z
  .enum(['it', 'es', 'fr', 'uk'])
  .nullish()
  .transform((x) => x ?? undefined)
  .default('it');
    
console.log({
  withValue: schema.safeParse('it'),
  withUndefined: schema.safeParse(undefined),
  withNull: schema.safeParse(null),
});

//     {
//       withValue: { success: true, data: 'it' },
//       withUndefined: { success: true, data: 'it' },
//       withNull: { success: true, data: undefined }
//     }
```

`null` value is successfully parsed but the result `data` is `undefined`.

I can use something like this: 

```ts
const schema = z
  .enum(['it', 'es', 'fr', 'uk'])
  .nullish()
  .transform((x) => x ?? 'it');
```

but I don't understand why the default function doesn't work after transformation.

**Comment by digisomni** [comment]
I think we need a built-in way to coerce it from "unknown" values.

**Reply by JacobWeisenburger** [reply]
What is the "it" that you are coercing from "unknown" values?

**Comment by movahhedi** [comment]
Something built-in like `.coerceUndefined` would be nice.

**Comment by skondrashov** [comment]
I would like something built-in because I'm trying to use this in thousands of places. All of our APIs provide `null` for optional fields, but I'd rather just deal with `undefined` in the code, so the simple and clean solution still doesn't feel clean enough.
